draft-ietf-quic-transport.md

```diff
@@ -499,41 +499,41 @@ The fields in a Regular packet past the Common Header are the following:
-CONNECTION_CLOSE frame with the error code QUIC_SEQUENCE_NUMBER_LIMIT_REACHED
-(connection termination is described in {{termination}}.)  For unambiguous
-reconstruction of the complete packet number by a receiver from the lower-order
-bits, a QUIC sender MUST NOT have more than 2^(packet_number_size - 2) in flight
-at any point in the connection.  In other words,
-
-* If a sender sets PACKET_NUMBER_SIZE bits to 11, it MUST NOT have more than
-  (2^46) packets in flight.
-
-* If a sender sets PACKET_NUMBER_SIZE bits to 10, it MUST NOT have more than
-  (2^30) packets in flight.
 
-* If a sender sets PACKET_NUMBER_SIZE bits to 01, it MUST NOT have more than
-  (2^14) packets in flight.
+## Packet Numbers
 
-* If a sender sets PACKET_NUMBER_SIZE bits to 00, it MUST NOT have more than
-  (2^6) packets in flight.
+The packet number is a 64-bit unsigned number and is used as part of a
+cryptographic nonce for packet encryption.  Each endpoint maintains a separate
+packet number for sending and receiving.  The packet number for sending MUST
+increase by one after sending any packet.
 
-  DISCUSS: Should the receiver be required to enforce this rule that the sender
-  MUST NOT exceed the inflight limit?  Specifically, should the receiver drop
-  packets that are received outside this window?
-
-  Any truncated packet number received from a peer MUST be reconstructed as the
-  value closest to the next expected packet number from that peer.
-
-(TODO: Clarify how packet number size can change mid-connection.)
+A QUIC endpoint MUST NOT reuse a packet number within the same connection (that
+is, under the same cryptographic keys).  If the packet number for sending
+reaches 2^64 - 1, the sender MUST close the connection by sending a
+CONNECTION_CLOSE frame with the error code QUIC_SEQUENCE_NUMBER_LIMIT_REACHED
+(connection termination is described in {{termination}}.)
+
+To reduce the number of bits required to represent the packet number over the
+wire, only the least significant bits of the packet number are transmitted over
+the wire, up to 48 bits.  The actual packet number for each packet is
+reconstructed at the receiver based on the largest packet number received on a
+successfully authenticated packet.
+
+A packet number is decoded by finding the packet number value that is closest to
+the next expected packet.  The next expected packet is the highest received
+packet number plus one.  For example, if the highest successfully authenticated
+packet had a packet number of 0xaa82f30e, then a packet containing a 16-bit
+value of 0x1f94 will be decoded as 0xaa831f94.
+
+To enable unambiguous reconstruction of the packet number, an endpoint MUST use
+a packet number size that is able to represent 4 times more packet numbers than
+the endpoint has currently outstanding.  A packet is outstanding if it sent but
+has neither been acknowledged nor been marked as lost (see {{QUIC-RECOVERY}}).
```

Comment:
Do we really want to define "outstanding" by the number of packets w/o ACKs?  I'd rather define it by the number of packet numbers between the current packet number and the earliest packet w/o ACKs.

For a high bandwidth, high latency connection, if I have packet 10 "delayed", but packets 11-1033 have been acked, do I really only have 1 packet "outstanding" and therefore can use the smallest number of bits to encode the following packet number?